ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -726,13 +726,20 @@ chainSyncClient mkPipelineDecision0 tracer cfg
     rollForward mkPipelineDecision n hdr theirTip
               = Stateful $ \kis -> traceException $ do
       now <- getMonotonicTime
-      let hdrHash  = headerHash hdr
-          hdrPoint = headerPoint hdr
-      unless (isPipeliningEnabled version) $ do
-        -- Reject the block if invalid
-        isInvalidBlock <- atomically $ forgetFingerprint <$> getIsInvalidBlock
-        whenJust (isInvalidBlock hdrHash) $ \reason ->
-          disconnect $ InvalidBlock hdrPoint reason
+      let hdrPoint = headerPoint hdr
+
+      isInvalidBlock <- atomically $ forgetFingerprint <$> getIsInvalidBlock
+      let disconnectWhenInvalid pt = case pointHash pt of
+            GenesisHash    -> pure ()
+            BlockHash hash ->
+              whenJust (isInvalidBlock hash) $ \reason ->
+                disconnect $ InvalidBlock pt reason
+      disconnectWhenInvalid $
+        if isPipeliningEnabled version
+        -- Disconnect if their tip is now known to be invalid. Usually, this is
+        -- the parent of hdr, which is also later checked explicitly.
+        then getTipPoint (unTheir theirTip)
```

Comment:
I think we want `headerPrevHash hdr`, not `theirTip`. `theirTip` is an (completely unchecked) `Point` that they send as a "helpful" annotation on their `MsgRollForward`.

Reply:
Thanks for catching this, I somehow assumed that this would be the tip of their fragment just before processing this `MsgRollForward`, but it is indeed this (currently unused apart from error msgs) bit we talked about previously.

I only ended up looking for other than `headerPrevHash hdr` because we need a `Point` for error reporting... Could we simply report only the hash instead of the point, or is there an easy way to get the point of `headerPrevHash hdr`?

Reply:
We could do this by moving the block validity check [here](https://github.com/input-output-hk/ouroboros-network/blob/6d11459940a5d9a06a0109b75e56bc37a03cf940/ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs#L797), but then we don't disconnect if the intersection check resulted in `NoLongerIntersects`.

Reply:
Hmmm. Actually, now that I'm following your lead, looking at these details, it occurs to me that `InvalidBlock` in some sense isn't what we should throw. Or at least, its semantics before pipelining are a bit different than after pipelining -- or at least as of this PR.

For example, I think the current expectation is that `InvalidBlock pt reason` is thrown when processing a `MsgRollForward` _that sent `pt`_. In our case, it's the `reason` that should specific the parent hash, because -- and this is the confusing part -- the parent block is invalid, which in turn means all blocks that extend it are also invalid.

Does that seem worthwhile to you too?

Reply:
Agreed, that seems sensible.

AFAIK, `InvalidBlockReason` does not contain the hash of the block (i.e. in Shelley, it seems to be `BlockTransitionError` from ledger on block body validation failure), which is why it is stored together with a `Point`/`RealPoint`/`Slot` in various places (`InvalidBlockInfo` in `ChainDB.Impl.Types`, or `TraceValidationEvent.InvalidBlock`, or `IgnoreInvalidBlock`, or `TraceForgedInvalidBlock`); so maybe the proper way to fix this would be to ensure that we can always extract a `Point` from `InvalidBlockReason`.

But I am not sure if changing this (especially in this PR) pays its weight, so I went with adding a `HeaderHash blk` argument to `InvalidBlock` for now.

Reply:
OK. After we just talked it through on the call for 5 minutes or so, I agree that just adding the hash to `InvalidBlock` is a reasonable small change to make here. The tension is that a whole new exception constructor seems excessive, while expressing this logic with a whole new constructor in `InvalidBlockReason` seems inappropriate, since ChainDB owns that type and will never itself emit that constructor.

Adding some more information to `InvalidBlock` as you've done here strikes a fine balance. There's still one constructor for  blocks (imagine grepping logs, eg), but it'll contain information to distinguish the cases (ie invalid block itself vs invalid parent)